price-feeder/oracle/provider/kraken.go

```diff
@@ -23,100 +26,391 @@ type (
+		return nil, err
+	}
+
+	go provider.handleWebSocketMsgs(ctx)
+
+	return provider, nil
+}
+
+// GetTickerPrices returns the tickerPrices based on the saved map.
+func (p *KrakenProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
+	tickerPrices := make(map[string]TickerPrice, len(pairs))
+
+	for _, cp := range pairs {
+		key := cp.String()
+		tickerPrice, ok := p.tickers[key]
+		if !ok {
+			return nil, fmt.Errorf("kraken provider failed to get ticker price for %s", key)
+		}
+		tickerPrices[key] = tickerPrice
+	}
+
+	return tickerPrices, nil
+}
+
+// SubscribeTickers subscribe to all currency pairs and
+// add the new ones into the provider subscribed pairs.
+func (p *KrakenProvider) SubscribeTickers(cps ...types.CurrencyPair) error {
+	pairs := make([]string, len(cps))
+
+	for i, cp := range cps {
+		pairs[i] = currencyPairToKrakenPair(cp)
+	}
+
+	if err := p.subscribePairs(pairs...); err != nil {
+		return err
 	}
+
+	p.setSubscribedPairs(cps...)
+	return nil
 }
 
-func NewKrakenProviderWithTimeout(timeout time.Duration) *KrakenProvider {
-	return &KrakenProvider{
-		baseURL: krakenBaseURL,
-		client:  newHTTPClientWithTimeout(timeout),
+// handleWebSocketMsgs receive all the messages from the provider
+// and controls to reconnect the web socket.
+func (p *KrakenProvider) handleWebSocketMsgs(ctx context.Context) {
+	reconnectTicker := time.NewTicker(defaultMaxConnectionTime)
+	defer reconnectTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewWSMessage):
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message
+				p.logger.Err(err).Msg("kraken provider could not read message")
+				if err := p.ping(); err != nil {
+					p.logger.Err(err).Msg("failed to send ping")
+					p.keepReconnecting()
+				}
+				continue
+			}
+
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.messageReceived(messageType, bz)
+
+		case <-reconnectTicker.C:
+			if err := p.reconnect(); err != nil {
+				p.logger.Err(err).Msg("kraken provider attempted to reconnect")
+				p.keepReconnecting()
+			}
+		}
 	}
 }
 
-func (p KrakenProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
-	tickers := make([]string, len(pairs))
-	for i, cp := range pairs {
-		tickers[i] = cp.String()
+// messageReceived handles any message sent by the provider.
+func (p *KrakenProvider) messageReceived(messageType int, bz []byte) {
+	if messageType != websocket.TextMessage {
+		return
 	}
 
-	path := fmt.Sprintf("%s%s?pair=%s", p.baseURL, krakenTickerEndpoint, strings.Join(tickers, ","))
+	var krakenEvent KrakenEvent
+	if err := json.Unmarshal(bz, &krakenEvent); err != nil {
+		p.logger.Debug().Msg("kraken provider received a message that is not an event")
+		// msg is not an event, it will try to marshal to ticker message
+		p.messageReceivedTickerPrice(bz)
```

Comment:
This is what you meant by the ugly parsing huh?

Reply:
yeah :/ 
Any idea how to make it better? 

Another bad thing is to [unmarshal](https://github.com/umee-network/umee/blob/843d8bfc22eff17ccbef10d39b7bdc3a5328b989/price-feeder/oracle/provider/kraken.go#L204) to an `[]Interface{}`